server/asset/eth/rpcclient.go

```diff
@@ -97,74 +105,122 @@ func (c *rpcclient) connect(ctx context.Context, log dex.Logger) error {
 		return fmt.Errorf("no contract address for eth version %d on %s", ethContractVersion, c.net)
 	}
 
-	es, err := swapv0.NewETHSwap(contractAddr, c.ec)
-	if err != nil {
-		return fmt.Errorf("unable to find swap contract: %v", err)
+	c.clients = make([]*ethConn, len(c.endpoints))
+	for i, endpoint := range c.endpoints {
+		client, err := rpc.DialContext(ctx, endpoint)
+		if err != nil {
+			return fmt.Errorf("unable to dial rpc to %q: %v", endpoint, err)
+		}
+
+		ec := &ethConn{
+			Client: ethclient.NewClient(client),
+			tokens: make(map[uint32]*tokener),
+		}
+
+		reqModules := []string{"eth", "txpool"}
+		if err := dexeth.CheckAPIModules(client, endpoint, c.log, reqModules); err != nil {
+			c.log.Warnf("Error checking required modules at %q: %v", endpoint, err)
+			c.log.Warnf("Will not account for pending transactions in balance calculations at %q", endpoint)
+			ec.txPoolSupported = false
+		} else {
+			ec.txPoolSupported = true
+		}
+
+		es, err := swapv0.NewETHSwap(contractAddr, ec.Client)
+		if err != nil {
+			return fmt.Errorf("unable to initialize eth contract for %q: %v", endpoint, err)
+		}
+		ec.swapContract = &swapSourceV0{es}
+		ec.caller = client
+
+		c.clients[i] = ec
 	}
-	c.swapContract = &swapSourceV0{es}
-	c.caller = client
 	return nil
 }
 
 // shutdown shuts down the client.
 func (c *rpcclient) shutdown() {
-	if c.ec != nil {
-		c.ec.Close()
+	for _, ec := range c.clients {
+		ec.Close()
 	}
 }
 
 func (c *rpcclient) loadToken(ctx context.Context, assetID uint32) error {
-	tkn, err := newTokener(ctx, assetID, c.net, c.ec)
-	if err != nil {
-		return fmt.Errorf("error constructing ERC20Swap: %w", err)
-	}
+	for _, cl := range c.clients {
+		tkn, err := newTokener(ctx, assetID, c.net, cl.Client)
+		if err != nil {
+			return fmt.Errorf("error constructing ERC20Swap: %w", err)
+		}
 
-	c.tokens[assetID] = tkn
+		cl.tokens[assetID] = tkn
+	}
 	return nil
 }
 
 func (c *rpcclient) withTokener(assetID uint32, f func(*tokener) error) error {
-	tkn, found := c.tokens[assetID]
-	if !found {
-		return fmt.Errorf("no swap source for asset %d", assetID)
-	}
-	return f(tkn)
+	return c.withClient(func(ec *ethConn) error {
+		tkn, found := ec.tokens[assetID]
+		if !found {
+			return fmt.Errorf("no swap source for asset %d", assetID)
+		}
+		return f(tkn)
+	})
+
 }
 
 // bestHeader gets the best header at the time of calling.
-func (c *rpcclient) bestHeader(ctx context.Context) (*types.Header, error) {
-	bn, err := c.ec.BlockNumber(ctx)
-	if err != nil {
-		return nil, err
-	}
-	return c.ec.HeaderByNumber(ctx, big.NewInt(int64(bn)))
+func (c *rpcclient) bestHeader(ctx context.Context) (hdr *types.Header, err error) {
+	return hdr, c.withClient(func(ec *ethConn) error {
+		bn, err := ec.BlockNumber(ctx)
+		if err != nil {
+			return err
+		}
+		hdr, err = ec.HeaderByNumber(ctx, big.NewInt(int64(bn)))
+		return err
```

Comment:
Doing some testing with Blast (blastapi.io) and I got a method not found, but I can't tell what method:

```
[ERR] ASSET[eth][RPC]: Unpropagated error from "wss://eth-mainnet.blastapi.io/xxxxx-xyyyy-adsfasdf": Method not found
[INF] ASSET[eth][RPC]: Switching RPC endpoint to "ws://127.0.0.1:8546"
```

On startup it did the balance method check and as expected it went the dumb way:

```
[DBG] ASSET[eth]: Parsed 2 endpoints from the ETH config file                                                                                                                      
[WRN] ASSET[eth][RPC]: Error checking required modules at "wss://eth-mainnet.blastapi.io/xxxxx-xyyyy-adsfasdf": unable to check supported modules: Method not found
[WRN] ASSET[eth][RPC]: Will not account for pending transactions in balance calculations at "wss://eth-mainnet.blastapi.io/xxxxx-xyyyy-adsfasdf"                   
[DBG] ASSET[eth][RPC]: API endpoints supported by ws://127.0.0.1:8546: admin:1.0 engine:1.0 eth:1.0 net:1.0 rpc:1.0 txpool:1.0          
```

So possibly there's another method missing and blast isn't gonna work, or the smart/dumb switch didn't work as expected.  I don't think it was checking a balance at the time, so it's probably another method sadly.

Reply:
You could add blastapi to your `~/dextest/providers.json` and run `go test -tags rpclive,lgpl -run TestMainnetCompliance` in **client/asset/eth** to see if anything pops out. I've seen `"Method not found"` for `SuggestGasTipCap` before. Not sure how that aligns with API availability.

Reply:
This turned out to be the `eth_syncing` method, which is expected (with https://github.com/decred/dcrdex/pull/2013/commits/06345a5ffcdcedfc1218a35ca8e58289892071bf).  Given this provider switching that happens as a result, I think I'll update that commit in #2013 to just not even try `eth_syncing` since it's next to useless anyway.

Reply:
#2013 revised with https://github.com/decred/dcrdex/pull/2013/commits/168cb574f8b128ea12b426b753e4261889ec5849
No more `eth_syncing` on client either.